numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -309,6 +314,79 @@ static int
 
 /**end repeat**/
 
+#ifdef HAVE_STRTOLD_L
+static NPY_INLINE npy_longdouble string_to_long_double(PyObject*op)
+{
+    char *s;
+    char *end;
+    npy_longdouble temp;
+    PyObject* b;
+
+    /* This weird structure is to make the error handling work:
+     * if it's not a string, we want to fall back to other
+     * approaches. In python 2, that happens when we try to
+     * convert the string to a C string. In python 3, we have
+     * to use a two-step conversion process, via bytes; the
+     * not-a-string situation gets triggered on conversion to 
+     * bytes, but we want to make sure bytes objects get 
+     * handled too, because python accepts bytes as input to
+     * float() (why?).
+     */
+    b = PyUnicode_AsUTF8String(op);
+    if (!b) {
+        /* Wasn't a string. Maybe it was a bytes? */
+        PyErr_Clear();
+        b = op;
+        Py_XINCREF(b);
+    }
+    s = PyBytes_AsString(b);
+
+    if (s) {
+        locale_t clocale;
+        errno = 0; 
+        clocale = newlocale(LC_ALL_MASK, "C", NULL);
+        if (!clocale) {
+            PyErr_Format(PyExc_ValueError,
+                         "could not convert string to long double: unable to set up C locale (%s)",
+                         strerror(errno));
+            Py_XDECREF(b);
+            return 0;
+        }
+
+        temp = strtold_l(s, &end, clocale);
+
+        freelocale(clocale);
+        
+        if (end==s || *end) {
+            PyErr_Format(PyExc_ValueError,
```

Comment:
I tend to break the style guide for these and just indent the following line 4 spaces to get more room for strings.